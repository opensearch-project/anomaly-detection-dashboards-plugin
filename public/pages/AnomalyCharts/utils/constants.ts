/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

export const INITIAL_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0.0,
  maxAnomalyGrade: 0.0,
  minConfidence: 0.0,
  maxConfidence: 0.0,
  lastAnomalyOccurrence: '',
};

export enum CHART_FIELDS {
  PLOT_TIME = 'plotTime',
  ANOMALY_GRADE = 'anomalyGrade',
  CONFIDENCE = 'confidence',
  DATA = 'data',
  AGG_INTERVAL = 'aggInterval',
  EXPECTED_VALUE = 'expectedValue',
}

export enum CHART_COLORS {
  ANOMALY_GRADE_COLOR = '#D13212',
  FEATURE_DATA_COLOR = '#16191F',
  FEATURE_COLOR = '#fcd529',
  CONFIDENCE_COLOR = '#017F75',
  LIGHT_BACKGROUND = '#FFFFFF',
  DARK_BACKGROUND = '#1D1E24',
}

export const FEATURE_CHART_THEME = [
  {
    lineSeriesStyle: {
      line: {
        strokeWidth: 2,
        visible: true,
        opacity: 0.5,
      },
      point: {
        visible: true,
        stroke: CHART_COLORS.FEATURE_DATA_COLOR,
      },
    },
  },
  {
    colors: {
      vizColors: [CHART_COLORS.FEATURE_DATA_COLOR],
    },
  },
];

export const ANOMALY_CHART_THEME = [
  {
    colors: {
      vizColors: [
        CHART_COLORS.CONFIDENCE_COLOR,
        CHART_COLORS.ANOMALY_GRADE_COLOR,
      ],
    },
  },
];

export const DATE_PICKER_QUICK_OPTIONS = [
  { start: 'now-24h', end: 'now', label: 'last 24 hours' },
  { start: 'now-7d', end: 'now', label: 'last 7 days' },
  { start: 'now-30d', end: 'now', label: 'last 30 days' },
  { start: 'now-90d', end: 'now', label: 'last 90 days' },

  { start: 'now/d', end: 'now', label: 'Today' },
  { start: 'now/w', end: 'now', label: 'Week to date' },
  { start: 'now/M', end: 'now', label: 'Month to date' },
  { start: 'now/y', end: 'now', label: 'Year to date' },
];

export enum LIVE_CHART_CONFIG {
  REFRESH_INTERVAL_IN_SECONDS = 30 * 1000, //poll anomaly result every 30 seconds
  MONITORING_INTERVALS = 60,
}

export const DEFAULT_ANOMALY_SUMMARY = {
  anomalyOccurrence: 0,
  minAnomalyGrade: 0,
  maxAnomalyGrade: 0,
  minConfidence: 0,
  maxConfidence: 0,
  lastAnomalyOccurrence: '-',
};

export const HEATMAP_CHART_Y_AXIS_WIDTH = 30;
