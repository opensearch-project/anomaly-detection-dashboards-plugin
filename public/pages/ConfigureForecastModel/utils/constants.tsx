/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import {
  ModelConfigurationFormikValues,
} from '../../ConfigureForecastModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

// FIXME: We intentionally leave some parameters undefined to encourage customers to click 
// the "Suggest parameters" button rather than providing default values like "10 minutes" for interval
export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined,
    interval: undefined,
    windowDelay: undefined,
    suggestedSeasonality: undefined,
    recencyEmphasis: undefined,
    resultIndex: undefined,
    resultIndexMinAge: 7,
    resultIndexMinSize: 51200,
    resultIndexTtl: 60,
    flattenCustomResultIndex: false,
  };

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
