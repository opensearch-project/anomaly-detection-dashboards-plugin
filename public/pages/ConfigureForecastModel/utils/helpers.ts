/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DATA_TYPES, DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';
import {
  Forecaster,
  UNITS,
} from '../../../models/interfaces';
import { get, cloneDeep, isEmpty } from 'lodash';
import {
  ModelConfigurationFormikValues,
  CustomValueFormikValues,
  ImputationFormikValues,
} from '../../ConfigureForecastModel/models/interfaces';
import { INITIAL_MODEL_CONFIGURATION_VALUES } from '../../ConfigureForecastModel/utils/constants';
import {
  ImputationMethod,
  ImputationOption,
} from '../../../models/types';
import {
  SparseDataOptionValue
} from './constants'

export const focusOnCategoryField = () => {
  const component = document.getElementById('categoryFieldCheckbox');
  component?.focus();
};

export const focusOnImputationOption = () => {
  const component = document.getElementById('imputationOption');
  component?.focus();
};

export const getShingleSizeFromObject = (obj: object) => {
  return get(obj, 'shingleSize', DEFAULT_SHINGLE_SIZE);
};

export function clearModelConfiguration(forecast: Forecaster): Forecaster {
  return {
    ...forecast,
    featureAttributes: [],
    uiMetadata: {
      ...forecast.uiMetadata,
      features: {},
    },
    categoryField: undefined,
    shingleSize: DEFAULT_SHINGLE_SIZE,
  };
}

export function createImputationFormikValues(
  forecaster: Forecaster
): ImputationFormikValues {
  const imputationMethod = imputationMethodToFormik(forecaster);
  let defaultFillArray: CustomValueFormikValues[] = [];

  if (SparseDataOptionValue.CUSTOM_VALUE === imputationMethod) {
    const defaultFill = get(forecaster, 'imputationOption.defaultFill', null) as Array<{ featureName: string; data: number }> | null;
    defaultFillArray = defaultFill
      ? defaultFill.map(({ featureName, data }) => ({
          featureName,
          data,
        }))
      : [];
  }

  return {
    imputationMethod: imputationMethod,
    custom_value: SparseDataOptionValue.CUSTOM_VALUE === imputationMethod ? defaultFillArray : undefined,
  };
}

export function modelConfigurationToFormik(
  forecaster: Forecaster
): ModelConfigurationFormikValues {
  const initialValues = cloneDeep(INITIAL_MODEL_CONFIGURATION_VALUES);
  if (isEmpty(forecaster)) {
    return initialValues;
  }

  const imputationFormikValues = createImputationFormikValues(forecaster);

  return {
    ...initialValues,
    shingleSize: get(forecaster, 'shingleSize', DEFAULT_SHINGLE_SIZE),
    imputationOption: imputationFormikValues,
    interval: get(forecaster, 'forecastInterval.period.interval', 10),
    windowDelay: get(forecaster, 'windowDelay.period.interval', 0),
    suggestedSeasonality: get(forecaster, 'suggestedSeasonality', undefined),
    recencyEmphasis: get(forecaster, 'recencyEmphasis', undefined),
    resultIndexMinAge: get(forecaster, 'resultIndexMinAge', undefined),
    resultIndexMinSize:get(forecaster, 'resultIndexMinSize', undefined),
    resultIndexTtl: get(forecaster, 'resultIndexTtl', undefined),
    flattenCustomResultIndex: get(forecaster, 'flattenCustomResultIndex', false),
    resultIndex: forecaster.resultIndex,
    horizon: get(forecaster, 'horizon', undefined),
    history: get(forecaster, 'history', undefined),
  };
}

export function formikToModelConfiguration(
  values: ModelConfigurationFormikValues,
  forecaster: Forecaster
): Forecaster {
  let forecasterBody = {
    ...forecaster,
    shingleSize: values.shingleSize,
    imputationOption: formikToImputationOption(
      forecaster.featureAttributes?.[0]?.featureName,
      values.imputationOption,
    ),
    forecastInterval: {
      period: { interval: values.interval, unit: UNITS.MINUTES },
    },
    windowDelay: {
      period: { interval: values.windowDelay, unit: UNITS.MINUTES },
    },
    resultIndexMinAge: values.resultIndexMinAge,
    resultIndexMinSize: values.resultIndexMinSize,
    resultIndexTtl: values.resultIndexTtl,
    flattenCustomResultIndex: values.flattenCustomResultIndex,
    resultIndex: values.resultIndex,
    horizon: values.horizon,
    history: values.history,
  } as Forecaster;

  return forecasterBody;
}

export function formikToImputationOption(
  featureName: string | undefined,
  imputationFormikValues?: ImputationFormikValues,
): ImputationOption | undefined {
  // If no feature name is provided, return undefined
  if (!featureName) return undefined;

  // Map the formik method to the imputation method; return undefined if method is not recognized.
  const method = formikToImputationMethod(imputationFormikValues?.imputationMethod);
  if (!method) return undefined;

  // Convert custom_value array to defaultFill if the method is FIXED_VALUES.
  const defaultFill = method === ImputationMethod.FIXED_VALUES
    ? imputationFormikValues?.custom_value?.map(({ data }) => ({
        featureName: featureName,
        data,
      }))
    : undefined;

  // Construct and return the ImputationOption object.
  return { method, defaultFill };
}

export function imputationMethodToFormik(
  forecaster: Forecaster
): string {
  var imputationMethod = get(forecaster, 'imputationOption.method', undefined) as ImputationMethod;

  switch (imputationMethod) {
    case ImputationMethod.FIXED_VALUES:
      return SparseDataOptionValue.CUSTOM_VALUE;
    case ImputationMethod.PREVIOUS:
      return SparseDataOptionValue.PREVIOUS_VALUE;
    case ImputationMethod.ZERO:
      return SparseDataOptionValue.SET_TO_ZERO;
    default:
      break;
  }

  return SparseDataOptionValue.IGNORE;
}

export function formikToImputationMethod(
  formikValue: string | undefined
): ImputationMethod | undefined {
  switch (formikValue) {
    case SparseDataOptionValue.CUSTOM_VALUE:
      return ImputationMethod.FIXED_VALUES;
    case SparseDataOptionValue.PREVIOUS_VALUE:
      return ImputationMethod.PREVIOUS;
    case SparseDataOptionValue.SET_TO_ZERO:
      return ImputationMethod.ZERO;
    default:
      return undefined;
  }
}

export const getCustomValueStrArray = (imputationMethodStr : string, forecaster: Forecaster): string[] => {
  if (SparseDataOptionValue.CUSTOM_VALUE === imputationMethodStr) {
    const defaultFill : Array<{ featureName: string; data: number }> = get(forecaster, 'imputationOption.defaultFill', []);

    return defaultFill
      .map(({ featureName, data }) => `${featureName}: ${data}`);
  }
  return []
}

export const toNumberOrEmpty = (value: string): number | '' => {
  return value === '' ? '' : Number(value);
};
