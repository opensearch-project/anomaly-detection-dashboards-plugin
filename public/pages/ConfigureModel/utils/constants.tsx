/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import {
  ModelConfigurationFormikValues,
  FeaturesFormikValues,
} from '../../ConfigureModel/models/interfaces';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';

export const INITIAL_MODEL_CONFIGURATION_VALUES: ModelConfigurationFormikValues =
  {
    featureList: [],
    categoryFieldEnabled: false,
    categoryField: [],
    shingleSize: DEFAULT_SHINGLE_SIZE,
    imputationOption: undefined
  };

export const INITIAL_FEATURE_VALUES: FeaturesFormikValues = {
  featureId: '',
  featureName: '',
  featureEnabled: true,
  featureType: FEATURE_TYPE.SIMPLE,
  aggregationQuery: JSON.stringify(
    { aggregation_name: { sum: { field: 'field_name' } } },
    null,
    4
  ),
  aggregationBy: '',
  aggregationOf: [],
};

export const FEATURE_TYPES = [
  { text: 'Custom Aggregation', value: FEATURE_TYPE.CUSTOM },
  { text: 'Defined Aggregation', value: FEATURE_TYPE.SIMPLE },
];

export const FEATURE_TYPE_OPTIONS = [
  { text: 'Field value', value: FEATURE_TYPE.SIMPLE },
  { text: 'Custom expression', value: FEATURE_TYPE.CUSTOM },
];

export enum SAVE_FEATURE_OPTIONS {
  START_AD_JOB = 'start_ad_job',
  KEEP_AD_JOB_STOPPED = 'keep_ad_job_stopped',
}

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'value_count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const FEATURE_FIELDS = [
  'featureName',
  'aggregationOf',
  'aggregationBy',
  'aggregationQuery',
];

// an enum for the sparse data handling options
export enum SparseDataOptionValue {
  IGNORE = 'ignore',
  PREVIOUS_VALUE = 'previous_value',
  SET_TO_ZERO = 'set_to_zero',
  CUSTOM_VALUE = 'custom_value',
}
