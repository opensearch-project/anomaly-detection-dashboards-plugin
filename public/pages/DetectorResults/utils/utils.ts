/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */


import { Detector } from '../../../models/interfaces';
import {
  NO_FULL_SHINGLE_ERROR_MESSAGE,
  NO_DATA_IN_WINDOW_ERROR_MESSAGE,
  NO_RCF_MODEL_ERROR_MESSAGE,
} from './constants';
import { DEFAULT_SHINGLE_SIZE } from '../../../utils/constants';
import { DETECTOR_STATE } from '../../../../server/utils/constants';
import moment, { Moment } from 'moment';
import { get } from 'lodash';
import { DETECTOR_INIT_FAILURES } from '../../../pages/DetectorDetail/utils/constants';

export const IS_INIT_OVERTIME_FIELD = 'isInitOvertime';
export const INIT_DETAILS_FIELD = 'initDetails';
export const INIT_ERROR_MESSAGE_FIELD = 'initErrorMessage';
export const INIT_ACTION_ITEM_FIELD = 'initActionItem';

export const getDetectorInitializationInfo = (detector: Detector) => {
  const currentTime = moment();

  let result = {
    [IS_INIT_OVERTIME_FIELD]: false,
    [INIT_DETAILS_FIELD]: {},
  };
  if (isDetectorInitOverTime(currentTime, detector)) {
    result[IS_INIT_OVERTIME_FIELD] = true;
    result[INIT_DETAILS_FIELD] = getInitOverTimeDetails(detector);
  }
  return result;
};

const isDetectorInitOverTime = (currentTime: Moment, detector: Detector) => {
  return (
    detector &&
    detector.curState === DETECTOR_STATE.INIT &&
    detector.stateError &&
    !detector.stateError.includes(NO_RCF_MODEL_ERROR_MESSAGE) &&
    //@ts-ignore
    currentTime
      .subtract(
        get(detector, 'shingleSize', DEFAULT_SHINGLE_SIZE) *
          detector.detectionInterval.period.interval,
        detector.detectionInterval.period.unit.toLowerCase()
      )
      //@ts-ignore
      .valueOf() > detector.enabledTime
  );
};

const getInitOverTimeDetails = (detector: Detector) => {
  let result = {
    [INIT_ERROR_MESSAGE_FIELD]: '',
    [INIT_ACTION_ITEM_FIELD]: '',
  };
  if (!detector.stateError) {
    return result;
  }
  if (detector.stateError.includes(NO_FULL_SHINGLE_ERROR_MESSAGE)) {
    result[INIT_ERROR_MESSAGE_FIELD] = 'of insufficient data';
    result[INIT_ACTION_ITEM_FIELD] =
      DETECTOR_INIT_FAILURES.NO_TRAINING_DATA.actionItem;
  } else if (detector.stateError.includes(NO_DATA_IN_WINDOW_ERROR_MESSAGE)) {
    result[INIT_ERROR_MESSAGE_FIELD] = 'no data could be found';
    result[INIT_ACTION_ITEM_FIELD] =
      'Make sure your source index has sufficient data in the current detector interval and try again.';
  }
  return result;
};
