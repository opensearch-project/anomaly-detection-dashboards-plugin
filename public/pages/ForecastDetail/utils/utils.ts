/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { Forecaster } from '../../../models/interfaces';
import { FORECASTER_STATE } from '../../../../server/utils/constants';

export const IS_INIT_OVERTIME_FIELD = 'isInitOvertime';
export const INIT_DETAILS_FIELD = 'initDetails';
export const INIT_ERROR_MESSAGE_FIELD = 'initErrorMessage';
export const INIT_ACTION_ITEM_FIELD = 'initActionItem';
export const DEFAULT_ENTITIES_TO_SHOW = 5;

export const getForecasterInitializationInfo = (forecaster: Forecaster) => {

  let result = {
    [IS_INIT_OVERTIME_FIELD]: false,
    [INIT_DETAILS_FIELD]: {},
  };
  if (isForecasterInitOverTime(forecaster)) {
    result[IS_INIT_OVERTIME_FIELD] = true;
    result[INIT_DETAILS_FIELD] = getInitOverTimeDetails(); 
  }
  return result;
};

const isForecasterInitOverTime = (forecaster: Forecaster) => {
  return (
    forecaster &&
    (forecaster.curState === FORECASTER_STATE.AWAITING_DATA_TO_INIT)
  );
};

const getInitOverTimeDetails = () => {
  return {
    [INIT_ERROR_MESSAGE_FIELD]: 'Insufficient data',
    [INIT_ACTION_ITEM_FIELD]: 'Ingest more data and restart the forecast. Or increase the forecast time interval and try again.',
  };
};
