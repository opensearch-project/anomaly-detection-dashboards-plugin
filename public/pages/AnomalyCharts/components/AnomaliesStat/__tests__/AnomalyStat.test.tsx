/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { render } from '@testing-library/react';
import {
  AnomalyStat,
  AnomalyStatWithTooltip,
  AlertsStat,
} from '../AnomalyStat';
import { getRandomMonitor } from '../../../../../redux/reducers/__tests__/utils';

describe('<AnomalyStat /> spec', () => {
  describe('Anomaly Stat', () => {
    test('renders component without tooltip', () => {
      const { container } = render(
        <AnomalyStat title="test title" description="test description" />
      );
      expect(container).toMatchSnapshot();
    });
  });
});

describe('<AnomalyStatWithTooltip /> spec', () => {
  describe('Anomaly Stat with tooltip', () => {
    test('renders component with tooltip and not loading', () => {
      const { container } = render(
        <AnomalyStatWithTooltip
          isLoading={false}
          minValue={0.5}
          maxValue={0.9}
          description="test description"
          tooltip="test tooltip"
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders component with tooltip and loading', () => {
      const { container } = render(
        <AnomalyStatWithTooltip
          isLoading={true}
          minValue={0.5}
          maxValue={0.9}
          description="test description"
          tooltip="test tooltip"
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});

describe('<AlertsStat /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('Alert Stat', () => {
    test('renders component with undefined monitor and loading', () => {
      console.error = jest.fn();
      const { container } = render(
        <AlertsStat
          monitor={undefined}
          showAlertsFlyout={jest.fn()}
          totalAlerts={9}
          isLoading={true}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders component with undefined monitor and not loading', () => {
      console.error = jest.fn();
      const { container } = render(
        <AlertsStat
          monitor={undefined}
          showAlertsFlyout={jest.fn()}
          totalAlerts={9}
          isLoading={false}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders component with monitor and not loading', () => {
      const { container } = render(
        <AlertsStat
          monitor={getRandomMonitor('test-detector-id', true)}
          showAlertsFlyout={jest.fn()}
          totalAlerts={9}
          isLoading={false}
        />
      );
      expect(container).toMatchSnapshot();
    });

    test('renders component with monitor and loading', () => {
      const { container } = render(
        <AlertsStat
          monitor={getRandomMonitor('test-detector-id', true)}
          showAlertsFlyout={jest.fn()}
          totalAlerts={9}
          isLoading={true}
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
