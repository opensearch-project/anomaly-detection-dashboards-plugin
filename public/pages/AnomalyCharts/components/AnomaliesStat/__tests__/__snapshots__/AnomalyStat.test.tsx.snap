// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertsStat /> spec Alert Stat renders component with monitor and loading 1`] = `
<div>
  <div
    class="euiStat euiStat--leftAligned"
  >
    <div
      class="euiText euiText--small euiStat__description"
    >
      <p>
        Alert
         
        <button
          class="euiLink euiLink--primary"
          style="font-size: 12px;"
          type="button"
        >
          Info
        </button>
      </p>
    </div>
    <div
      class="euiTitle euiTitle--small euiStat__title"
      style="display: inline;"
    >
      <p
        class="euiTitle euiTitle--small euiStat__title"
        style="display: inline;"
      >
        -
         
      </p>
      <a
        class="euiLink euiLink--primary"
        href="/1d42b75b-3f59-550a-a"
        rel="noopener noreferrer"
        style="font-size: 14px;"
        target="_blank"
      >
        View monitor
        <svg
          aria-label="External link"
          class="euiIcon euiIcon--small euiLink__externalIcon"
          focusable="false"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13 8.5a.5.5 0 1 1 1 0V12a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h3.5a.5.5 0 0 1 0 1H4a1 1 0 0 0-1 1v8a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1V8.5Zm-5.12.339a.5.5 0 1 1-.706-.707L13.305 2H10.5a.5.5 0 1 1 0-1H14a1 1 0 0 1 1 1v3.5a.5.5 0 1 1-1 0V2.72L7.88 8.838Z"
          />
        </svg>
        <span
          class="euiScreenReaderOnly"
        >
          (opens in a new tab or window)
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`<AlertsStat /> spec Alert Stat renders component with monitor and not loading 1`] = `
<div>
  <div
    class="euiStat euiStat--leftAligned"
  >
    <div
      class="euiText euiText--small euiStat__description"
    >
      <p>
        Alert
         
        <button
          class="euiLink euiLink--primary"
          style="font-size: 12px;"
          type="button"
        >
          Info
        </button>
      </p>
    </div>
    <div
      class="euiTitle euiTitle--small euiStat__title"
      style="display: inline;"
    >
      <p
        class="euiTitle euiTitle--small euiStat__title"
        style="display: inline;"
      >
        9
         
      </p>
      <a
        class="euiLink euiLink--primary"
        href="/899c56c6-99d0-539b-8"
        rel="noopener noreferrer"
        style="font-size: 14px;"
        target="_blank"
      >
        View monitor
        <svg
          aria-hidden="true"
          aria-label="External link"
          class="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
          focusable="false"
          height="16"
          role="img"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
          />
        </svg>
        <span
          class="euiScreenReaderOnly"
        >
          (opens in a new tab or window)
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`<AlertsStat /> spec Alert Stat renders component with undefined monitor and loading 1`] = `
<div>
  <div
    class="euiStat euiStat--leftAligned"
  >
    <div
      class="euiText euiText--small euiStat__description"
    >
      <p>
        Alert
         
        <button
          class="euiLink euiLink--primary"
          style="font-size: 12px;"
          type="button"
        >
          Info
        </button>
      </p>
    </div>
    <div
      class="euiTitle euiTitle--small euiStat__title"
      style="display: inline;"
    >
      <p
        class="euiTitle euiTitle--small euiStat__title"
        style="display: inline;"
      >
        -
         
      </p>
    </div>
  </div>
</div>
`;

exports[`<AlertsStat /> spec Alert Stat renders component with undefined monitor and not loading 1`] = `
<div>
  <div
    class="euiStat euiStat--leftAligned"
  >
    <div
      class="euiText euiText--small euiStat__description"
    >
      <p>
        Alert
         
        <button
          class="euiLink euiLink--primary"
          style="font-size: 12px;"
          type="button"
        >
          Info
        </button>
      </p>
    </div>
    <div
      class="euiTitle euiTitle--small euiStat__title"
      style="display: inline;"
    >
      <p
        class="euiTitle euiTitle--small euiStat__title"
        style="display: inline;"
      >
        9
         
      </p>
    </div>
  </div>
</div>
`;

exports[`<AnomalyStat /> spec Anomaly Stat renders component without tooltip 1`] = `
<div>
  <div
    class="euiStat euiStat--leftAligned"
  >
    <div
      class="euiText euiText--small euiStat__description"
    >
      test description
    </div>
    <div
      class="euiTitle euiTitle--small euiStat__title"
      style="display: inline;"
    >
      test title
    </div>
  </div>
</div>
`;

exports[`<AnomalyStatWithTooltip /> spec Anomaly Stat with tooltip renders component with tooltip and loading 1`] = `
<div>
  <div
    class="euiStat euiStat--leftAligned"
  >
    <div
      class="euiText euiText--small euiStat__description"
    >
      <p>
        test description
         
        <span
          class="euiToolTipAnchor"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7.5 11.508 7.468 8H6.25V7h2.401l.03 3.508H9.8v1H7.5Zm-.25-6.202a.83.83 0 0 1 .207-.577c.137-.153.334-.229.59-.229.256 0 .454.076.594.23.14.152.209.345.209.576 0 .228-.07.417-.21.568-.14.15-.337.226-.593.226-.256 0-.453-.075-.59-.226a.81.81 0 0 1-.207-.568ZM8 13A5 5 0 1 0 8 3a5 5 0 0 0 0 10Zm0 1A6 6 0 1 1 8 2a6 6 0 0 1 0 12Z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </p>
    </div>
    <div
      class="euiTitle euiTitle--small euiStat__title"
      style="display: inline;"
    >
      -
    </div>
  </div>
</div>
`;

exports[`<AnomalyStatWithTooltip /> spec Anomaly Stat with tooltip renders component with tooltip and not loading 1`] = `
<div>
  <div
    class="euiStat euiStat--leftAligned"
  >
    <div
      class="euiText euiText--small euiStat__description"
    >
      <p>
        test description
         
        <span
          class="euiToolTipAnchor"
        >
          <svg
            aria-hidden="true"
            class="euiIcon euiIcon--medium euiIcon-isLoading"
            focusable="false"
            height="16"
            role="img"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
            />
          </svg>
        </span>
      </p>
    </div>
    <div
      class="euiTitle euiTitle--small euiStat__title"
      style="display: inline;"
    >
      0.50-0.90
    </div>
  </div>
</div>
`;
